app/models/rubygem.rb

```diff
@@ -261,6 +261,62 @@ def first_built_date
     versions.by_earliest_built_at.limit(1).last.built_at
   end
 
+  def self.compact_index_versions(date)
+    created_gems = Rubygem.joins(:versions)
+      .select("name, versions.created_at as date, number, platform, info_checksum")
+      .where("versions.created_at > ?", date)
+      .order("versions.created_at, versions.number, platform")
+
+    yanked_gems = Rubygem.joins(:versions)
+      .select("name, versions.yanked_at as date, '-'||number, platform, info_checksum")
+      .where("indexed is false and versions.created_at > ?", date)
+      .order("versions.created_at, versions.number, platform")
+
+    # Arel don't support yet order on unions :(
+    # https://github.com/rails/arel/issues/98
+    sql = "(#{created_gems.to_sql}) UNION (#{yanked_gems.to_sql}) ORDER BY date"
+    gems = Rubygem.find_by_sql(sql)
+
+    gems.map do |gem|
+      CompactIndex::Gem.new(gem.name, [
+                              CompactIndex::GemVersion.new(
+                                gem.number,
+                                gem.platform,
+                                gem.info_checksum
+                              )
+                            ])
+    end
+  end
+
+  def compact_index_info
+    group_by_columns =
+      "number, platform, sha256, info_checksum, required_ruby_version, required_rubygems_version, versions.created_at"
+    dep_req_agg =
+      "string_agg(dependencies.requirements, '@' order by rubygems_dependencies.name)"
+    dep_name_agg =
+      "string_agg(rubygems_dependencies.name, ',' order by rubygems_dependencies.name) as dep_name"
+
+    result = Rubygem.includes(versions: { dependencies: :rubygem })
+      .where("rubygems.name = ? and indexed = true and (scope = 'runtime' or scope is null)", name)
+      .group(group_by_columns)
+      .order("versions.created_at, number, platform, dep_name")
+      .pluck("#{group_by_columns}, #{dep_req_agg}, #{dep_name_agg}")
```

Comment:
So this query contains:
```
LEFT OUTER JOIN "rubygems" "rubygems_dependencies" ON "rubygems_dependencies"."id" = "dependencies"."rubygem_id"
```
Join in `"rubygems_dependencies"."id" = "dependencies"."rubygem_id"` is a problem because if dependency has a gem which is not registered with rg.org then its `rubygem_id` entry will be empty.
```sql
 id  | requirements | rubygem_id | version_id | unresolved_name 
-----+--------------+------------+------------+-----------------
 782 | ~> 10.0      |            |        156 | does_not_exist
 781 | ~> 1.0       |         41 |        156 | # this gem exists hence has rubygem_id
 780 | ~> 1.0       |            |        156 | rake #this one has space after rake in gemspec "rake "
```

Here we will have mismatch for number of requirement versions and number of dep_names. This was not a problem on bundler-api because it added dep only if it had a name:
```Ruby
deps[key] << [row[:dep_name], row[:requirements]] if row[:dep_name]
```

So what are we doing here? Skipping if it dep is not registered with rg.org or using `unresolved_name` (will complicate things).